data/bash-completion/fwupdtool

```diff
@@ -76,6 +76,15 @@ _show_firmware_types()
 	COMPREPLY+=( $(compgen -W "${firmware_types}" -- "$cur") )
 }
 
+_show_device_ids()
+{
+	if ! command -v jq &> /dev/null; then
+		return 0
+	fi
+	local description
+	description="$(command fwupdmgr get-devices --json 2>/dev/null | jq '.Devices | .[] | .DeviceId')"
```

Comment:
I don't think fwupdtool can bash autocomplete with fwupdmgr can it?

Reply:
In general I don't think we want to be calling one from the other if we can avoid it.

Reply:
The reason I'm saying to avoid it is that fwupdtool will close fwupd daemon when it runs.  However fwupdtool is a lot slower to startup, particularly for bash completion.

Maybe we should make an exception on this.  @hughsie thoughts?

Reply:
I think two things fighting for the same lock is a terrible idea. Maybe fwupd could write a list of current device Ids somewhere in /var/lib/fwupd for just this purpose?

Reply:
Yeah that's a good idea!  Don't we do that with fwupdtool?  I seem to remember there was an argument to make it happen in chrome.

Reply:
> The reason I'm saying to avoid it is that fwupdtool will close fwupd daemon when it runs. However fwupdtool is a lot slower to startup, particularly for bash completion.
> 

Oh, that is not kind indeed... `fwupdmgr attach [TAB]` would start the daemon to list the devices-id, and kill it then (if is not up and running already).

I preferred used `fwupdmgr` because if is faster to get the device-ids cached by the daemon; `fwupdtool get-devices` may take a while and more to output the device-ids.

For now, it makes sense to make the change, with `fwupdtool` instead. However, `fwupdtool` would kill the daemon anyway. And the `fwupdmgr` bash completion script should not run `fwupdtool` for completion as it is like doing hara-kiri (killing the daemon to restart it just for completion).

I need time for reflexion...

Reply:
> Yeah that's a good idea! Don't we do that with fwupdtool? I seem to remember there was an argument to make it happen in chrome.

So we have a `--enable-json-state` argument for `fwupdtool`.  I think this would make sense to use to store the devices.  
I would propose that we drop/deprecate that option and add a configuration file option for that behavior (on by default for `fwupdtool`).  Then any time that `fwupdtool` is launched it updates the state file and your PR uses that for bash completion invisibly.  If it's missing, then you return 0.  You launch get-devices once it updates.

> Btw, why do we have two tools already? I am laking a bit of history.

Ah history.  So originally we have only client / daemon architecture but then there are certain commands/arguments that don't make sense to expose to all unprivileged users.  For example `install-blob`.  These were put into a tool that wasn't in PATH for advanced use cases.  Later it was determined that we shouldn't put tools like that out of PATH, and it got moved to PATH.  
Later again there was a desire to adopt fwupd in Chrome OS but they weren't willing to take on a new privileged daemon on the system.  So it was overhauled to support many of the same features as the client does.

Later that changed and now ChromeOS takes client/daemon.

But the net outcome is that now we have two tools with lots of overlapping functionality.  `fwupdtool` is more useful for developers/developing plugins and `fwupdmgr` is more useful for end users who do things like update their firmware only.

Hopefully I remembered that all right :)

Reply:
Makes sense to me. Probably /var/cache though.

Reply:
@hughsie I am running out of time too. I would do it later.

Reply:
When/where should I create the file? in the `fu_engine_update_motd()`?

Reply:
Maybe not that exact function, but yes, another new function that's called at the same time as that one.

Reply:
I have pushed an initial sketch.

However, it does not update the file if a device is un-pluged; any idea?

_Unrelated_: I have noticed that removing the device causes the wait for the re-plug delay, and I guess this is not useful in case of none-upgrade unless I miss something.

```
08:44:54:0472 FuEngine             auto-removing backend device Rival 3 Wireless [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
08:44:54:0472 FuDeviceList         waiting 40000ms for Rival 3 Wireless device removal
(...)
08:45:34:0485 FuDeviceList         doing delayed removal
08:45:34:0486 FuDeviceList         ::removed b0a78eb71f4eeea7df8fb114522556ba8ce22074
```

Reply:
Is fu_engine_emit_changed() on remove? It might only be fired after the 40000ms delayed removal.

> Unrelated: I have noticed that removing the device causes the wait for the re-plug delay

Out of interest; if you pull in the patch from wip/hughsie/ONLY_WFR_REMOVE_DELAY does that make things more obvious? You'll have to set FU_DEVICE_INTERNAL_FLAG_ONLY_WFR_REMOVE_DELAY on the device in _init().

Reply:
```
09:33:50:0350 FuEngine             auto-removing backend device Rival 3 Wireless [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
09:33:50:0350 FuDeviceList         ::removed b0a78eb71f4eeea7df8fb114522556ba8ce22074
```
:tada:

Reply:
And tested after `_write()`

```
09:37:04:0367 FuDeviceList         waiting 40000ms for replug
09:37:26:0430 FuDevice             already set GType to FuSteelseriesFizz, ignoring FuSteelseriesFizz
09:37:26:0431 FuDevice             using e8055e4e2470749d3ebc7735868bf869c94ebd8f for usb:03:00:02:03
09:37:26:0435 FuPlugin             emit added from steelseries: e8055e4e2470749d3ebc7735868bf869c94ebd8f
09:37:26:0435 FuDeviceList         ::added e8055e4e2470749d3ebc7735868bf869c94ebd8f
09:37:26:0436 FuEngine             writing motd target /home/gportay/src/GOO0034/fwupd/cache/motd.d/85-fwupd
09:37:26:0437 FuCommon             writing /home/gportay/src/GOO0034/fwupd/cache/devices.json with 1837 bytes
09:37:44:0368 FuHistory            modifying device Rival 3 Wireless [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
09:37:44:0383 FuHistory            modifying device Rival 3 Wireless [b0a78eb71f4eeea7df8fb114522556ba8ce22074]
09:37:44:0391 FuIdle               uninhibiting: update
```

Reply:
> Is fu_engine_emit_changed() on remove? It might only be fired after the 40000ms delayed removal.

I added `fu_engine_emit_changed()` on remove.

Reply:
I think we need to jq that cached file now, right?

Reply:
With redirecting `stderr` to `null` to hide error if file does not exist.

Reply:
@hughsie I am not a master or `meson`; what is the best solution to `sed`-ish the bash completion script if the prefix is changed at configure time? because the path to the file is hardcoded for now.

However, this is pointless if I am parsing the file in `fwupdtool` if using a new option (`--use-cache`?) or reusing `--enable-json-state` mentioned by @superm1.

Reply:
That's done with configure_file() e.g. see https://github.com/fwupd/fwupd/blob/main/data/installed-tests/meson.build -- the filename doesn't have to have a .in suffix, that's a bit of a relic from the past we don't need to repeat.